src/v/storage/segment_appender.h

```diff
@@ -162,21 +205,76 @@ class segment_appender {
     // still heavy weight operations compared to regular flush()
     ss::future<> hard_flush();
 
+    enum write_state { QUEUED = 1, DISPATCHED, DONE };
+
     struct inflight_write {
-        bool done;
-        size_t offset;
+        // true if the write extends to the end of the chunk, i.e., this
+        // is the last write that will use the current chunk before it
+        // is recycled
+        bool full;
+
+        // the current state of the write
+        write_state state = QUEUED;
+
+        // the chunk containing the data for this write, may be nulled out
+        // when the chunk moves to DONE state
+        chunk_ptr chunk;
+
+        // the aligned begin and end offsets into the chunk covering the
+        // region to write
+        size_t chunk_begin, chunk_end;
+
+        // The aligned file offset where this write begins. That is, the bytes
+        // at chunk.data() + chunk_begin are written to file_start_offset in the
+        // file.
+        size_t file_start_offset;
+
+        // The committed file offset after this this write, i.e., the offset
+        // one beyond the last byte logically written by this write.
+        size_t committed_offset;
+
+        /**
+         * @brief Set the state of the write
+         *
+         * I'm just here as a handy place to hide an assert.
+         */
+        void set_state(write_state new_state) {
+            // the only allowed transitions are QUEUED -> DISPATCHED -> DONE
+            vassert(
+              (state == QUEUED || state == DISPATCHED)
+                && new_state == state + 1,
+              "bad transition {} -> {}",
+              state,
+              new_state);
+            state = new_state;
+        }
 
-        explicit inflight_write(size_t offset)
-          : done(false)
-          , offset(offset) {}
+        /**
+         * @brief Try to merge the given write with this one.
+         *
+         * If the passed write is "compatible" with this write, merge
+         * it with this one so both writes are covered by "this" object.
+         *
+         * This is not a general purpose method: it expects to only be passed
+         * the immediately subsequent write as "other" and so asserts on
+         * conditions that must be true in this case rather than simply checking
+         * them as part of the merge check.
+         *
+         * @param other the write to try to merge in to this one
+         * @param prior_committed_offset the offset prior to updating the offset
+         * for the 'other' write, used only as a sanity check that the writes
+         * are adjacent.
+         * @return true iff the merge succeeded and this object was updated
+         */
+        bool
+        try_merge(const inflight_write& other, size_t prior_committed_offset);
```

Comment:
The crux of this change is that we will call `try_merge` on the last in-flight write in the dbhw path.